winpr/libwinpr/clipboard/dirent.c

```diff
@@ -0,0 +1,91 @@
+
+#include <stdio.h>
+#include "dirent.h"
+
+#include "../log.h"
+#define TAG WINPR_TAG("clipboard.synthetic.file")
+
+DIR* opendir(const char* name)
+{
+	DIR* dir = NULL;
+	WIN32_FIND_DATAW FindData;
+	char namebuf[MAX_PATH];
+	WCHAR* file = NULL;
+	int nLen = 0;
+	int nRet = 0;
+
+	sprintf(namebuf, "%s\\*.*", name);
+	nLen = strlen(namebuf);
+
+	do
+	{
+		nRet = ConvertToUnicode(CP_UTF8, 0, namebuf, -1, file, 0);
+		if (nRet != nLen)
+			break;
+
+		HANDLE hFind = FindFirstFileW(file, &FindData);
+		if (hFind == INVALID_HANDLE_VALUE)
+		{
+			WLog_ERR(TAG, "FindFirstFile failed (%d)\n", GetLastError());
+			break;
+		}
+
+		dir = (DIR*)calloc(1, sizeof(DIR));
+		if (!dir)
+		{
+			WLog_ERR(TAG, "DIR memory allocate fail\n");
+			break;
+		}
+
+		dir->dd_fd = hFind; // simulate return
+	} while (0);
+
+	if (file)
```

Comment:
`free(ptr)` does work with `NULL` pointers, no need to check.

Reply:
Yes. Do a pointer check, keep it a good habit.

Reply:
@KangLin actually checking for `NULL` before calling `free` is bad practice as `free(NULL)` is well defined by the standard and considered safe.

Reply:
@mfleisz  This function does not need to check this pointer and will not make an error, but not checking the pointer before calling other functions may cause the program to crash.

Reply:
@KangLin while this is true for many functions `free(NULL)` is `well defined` by the standard, so in case there is a crash fix your `libc` because it is broken ;)